File: Android Project/app/src/main/java/com/example/histoquiz/activities/MenuActivity.java
package com.example.histoquiz.activities;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.view.View;
import android.widget.Button;
import com.example.histoquiz.R;

/**
 * Classe utilizada para manipular os cliques recebidos na tela de menu do jogo, redirecionando o
 * usuário para as activities corretas relativas a cada interação
 */
public class MenuActivity extends AppCompatActivity implements View.OnClickListener {

    protected Button button1, button2, button3, button4, voltar;

    // Variáveis para o controle da tela como fullscreen
    private final Handler mHideHandler = new Handler();
    private View mContentView;


    /**
     * Método chamado assim que essa activity é invocada.
     * @param savedInstanceState - contém o estado anteriormente salvo da atividade (pode ser nulo)
     */
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_menu);
        initGui();
        setInicialMenuOptions();
        mContentView = findViewById(R.id.fullContent);
        hideNow();
    }


    /**
     * Runnable utilizado para remover automaticamente a barra de botões e a de status dessa
     * activity
     */
    private final Runnable mHidePart2Runnable = new Runnable() {
        @SuppressLint("InlinedApi")
        @Override
        public void run() {
            mContentView.setSystemUiVisibility(View.SYSTEM_UI_FLAG_LOW_PROFILE
                    | View.SYSTEM_UI_FLAG_FULLSCREEN
                    | View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                    | View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
                    | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                    | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION);
        }
    };


    /**
     * Runnable utilizado para exibir a barra de botões e a de status dessa activity quando o
     * usuário solicitar
     */
    private final Runnable mShowPart2Runnable = () -> {
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.show();
        }
    };


    /**
     * Programa uma chamada ao método hide() após uma quantidade delayMillis de millisegundos,
     * cancelando qualquer chamada programada previamente
     */
    public void hideNow() {
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.hide();
        }
        mHideHandler.removeCallbacks(mShowPart2Runnable);
        mHideHandler.postDelayed(mHidePart2Runnable, 0);
    }


    /** Método utilizado para obter uma referência para os quatro botões presentes na view
     * relacionada a essa activity
     */
    protected void initGui(){
        button1 = findViewById(R.id.button1);
        button1.setOnClickListener(this);
        button2 = findViewById(R.id.button2);
        button2.setOnClickListener(this);
        button3 = findViewById(R.id.button3);
        button3.setOnClickListener(this);
        button4 = findViewById(R.id.button4);
        button4.setOnClickListener(this);
        voltar = findViewById(R.id.voltar);
        voltar.setOnClickListener(this);
    }


    /**
     * Método utilizado para redirecionar os cliques recebidos nos botões da view relacionada
     * a essa activity para o método que lidará corretamente com ele
     * @param v - view onde o clique ocorreu
     */
    @Override
    public void onClick(View v) {
        String tag = (String) v.getTag();
        switch (tag){
            case "MINHA_CONTA":            myAccount();            break;
            case "JOGAR":                  setPlayOptions();       break;
            case "REVISAR":                setReviewOptions();     break;
            case "SOBRE_O_JOGO":           setAboutOptions();      break;
            case "JOGAR_ONLINE":           setPlayOnlineOptions(); break;
            case "JOGAR_PC":               playAgainstPC();        break;
            case "CONVIDAR_AMIGO":         inviteFriend();         break;
            case "MODO_ALEATORIO":         randomMode();           break;
            case "O_QUE_E":                whatItIs();             break;
            case "COMO_JOGAR":             howToPlay();            break;
            case "FALE_CONOSCO":           contactUs();            break;
            case "SISTEMA_REPRODUTOR":     showReview(1);     break;
            case "SISTEMA_DIGESTORIO":     showReview(2);     break;
            case "SISTEMA_RESPIRATORIO":   showReview(3);     break;
            case "SISTEMA_CARDIOVASCULAR": showReview(4);     break;
            case "MENU_INICIAL":           setInicialMenuOptions();break;
        }
    }


    /**
     * Método utilizado para trocar para a activity responsável por exibir os
     * detalhes da conta do usuário
     */
    protected void myAccount(){
        Intent troca = new Intent(MenuActivity.this, MyAccountActivity.class);
        startActivity(troca);
    }


    /**
     * Método que exibe ao jogador as opções de jogo disponíveis: jogar contra
     * o computador ou jogar online. Para isso, ele se aproveita dos botões do
     * menu: "jogar" vira "jogar online" e "revisar" vira "jogar contra o com-
     * putador". Os demais botões são desabilitados e ficam invisíveis
     */
    protected void setPlayOptions(){
        button1.setVisibility(View.GONE);
        button1.setEnabled(false);
        button4.setVisibility(View.INVISIBLE);
        button4.setEnabled(false);
        button2.setText(R.string.jogarOnline);
        button2.setTag("JOGAR_ONLINE");
        button3.setText(R.string.jogarPC);
        button3.setTag("JOGAR_PC");
        voltar.setVisibility(View.VISIBLE);
        voltar.setEnabled(true);
        voltar.setTag("MENU_INICIAL");
    }


    /**
     * Método utilizado para redirecionar dessa activity para a que exibe
     */
    protected void setReviewOptions(){
        button1.setText(R.string.sistemaReprodutor);
        button1.setTag("SISTEMA_REPRODUTOR");
        button2.setText(R.string.sistemaDigestorio);
        button2.setTag("SISTEMA_DIGESTORIO");
        button3.setText(R.string.sistemaCardiopulmonar);
        button3.setTag("SISTEMA_RESPIRATORIO");
        button4.setText(R.string.sistemaOsteomuscular);
        button4.setTag("SISTEMA_CARDIOVASCULAR");
        voltar.setVisibility(View.VISIBLE);
        voltar.setEnabled(true);
        voltar.setTag("MENU_INICIAL");
    }


    /**
     * Método que exibe ao jogador as opções sobre o jogo: o que é, como jogar e fale co-
     * nosco. Para isso, ele se aproveita dos botões do menu: "jogar", transformando-o em
     * "O que é"; "revisar", transformando-o em "como jogar"; e "sobre o jogo", transfor-
     * mando-o em "fale conosco"
     */
    private void setAboutOptions(){
        button1.setVisibility(View.GONE);
        button2.setText(R.string.oQue);
        button2.setTag("O_QUE_E");
        button3.setText(R.string.comoJogar);
        button3.setTag("COMO_JOGAR");
        button4.setText(R.string.faleConosco);
        button4.setTag("FALE_CONOSCO");
        voltar.setVisibility(View.VISIBLE);
        voltar.setEnabled(true);
        voltar.setTag("MENU_INICIAL");
    }


    /**
     * Método que exibe ao jogador as opções de jogo online disponíveis: jogar contra
     * um amigo ou jogar no modo aleatório (o jogo escolhe um oponente qualquer). Para
     * isso, ele se aproveita dos botões do menu: "jogar"(que já tinha se tornado "jo-
     * gar online"), transformando-o em "Convidar amigo", e "revisar" (que já tinha se
     * tornado "jogar contra o computador"), transformando-o em "modo aleatório"
     */
    private void setPlayOnlineOptions(){
        button2.setText(R.string.convidarAmigo);
        button2.setTag("CONVIDAR_AMIGO");
        button3.setText(R.string.modoAleatorio);
        button3.setTag("MODO_ALEATORIO");
        voltar.setVisibility(View.VISIBLE);
        voltar.setEnabled(true);
        voltar.setTag("JOGAR");
    }


    /**
     * Método utilizado para redirecionar o jogador diretamente para a tela de jogo,
     * já que ele irá jogar contra o próprio celular
     */
    protected void playAgainstPC(){
        Intent troca = new Intent(MenuActivity.this, GameActivity.class);
        troca.putExtra("matchCreator", true);
        troca.putExtra("PCopponent", true);
        troca.putExtra("opponentUID", "0");
        startActivityForResult(troca, 999);
    }

    /**
     * Método utilizado para redirecionar dessa activity para a que exibe a tela
     * utilizada para convidar amigo para jogar
     */
    private void inviteFriend(){
        Intent troca = new Intent(MenuActivity.this, InviteFriendToPlayActivity.class);
        startActivity(troca);
    }


    /**
     * Método utilizado para redirecionar dessa activity para a de jogo, após encon-
     * trar um oponente aleatório para jogar contra esse usuário
     */
    private void randomMode(){

    }


    /**
     * Método utilizado para redirecionar dessa activity para a que descreve
     * o que o HistoQuiz é
     */
    private void whatItIs(){

    }

    /**
     * Método utilizado para redirecionar dessa activity para a que descreve
     * como jogar o HistoQuiz
     */
    private void howToPlay(){

    }


    /**
     * Método utilizado para redirecionar dessa activity para a que permite
     * ao usuário enviar um email para os desenvolvedores do HistoQuiz
     */
    private void contactUs(){

    }


    /**
     * Método utilizado para redirecionar dessa activity para a que exibe ao usuário
     * os dados para ele revisar sobre o sistema escolhido
     * @param kind - sistema do corpo humano que o usuário deseja revisar, a saber:
     *               1 - Sistema reprodutor
     *               2 - Sistema digestório
     *               3 - Sistema respiratório
     *               4 - Sistema cardiovascular
     */
    private void showReview(int kind){

    }


    /**
     * Método utilizado para setar tags para os botões presentes na view relacionada
     * a essa activity para a sua primeira execução, que exibe o menu principal
     */
    protected void setInicialMenuOptions(){
        button1.setText(R.string.minhaConta);
        button1.setTag("MINHA_CONTA");
        button1.setVisibility(View.VISIBLE);
        button1.setEnabled(true);
        button2.setText(R.string.jogar);
        button2.setTag("JOGAR");
        button2.setVisibility(View.VISIBLE);
        button2.setEnabled(true);
        button3.setText(R.string.revisar);
        button3.setTag("REVISAR");
        button3.setVisibility(View.VISIBLE);
        button3.setEnabled(true);
        button4.setText(R.string.sobreJogo);
        button4.setTag("SOBRE_O_JOGO");
        button4.setVisibility(View.VISIBLE);
        button4.setEnabled(true);
        voltar.setVisibility(View.INVISIBLE);
        voltar.setEnabled(false);
    }


    /**
     * Sobrescreve o método que define o que deve ser feito quando o botão
     * pressionar do celular for pressionado. Como não quero que ele seja
     * utilizado, o método fica vazio
     */
    @Override
    public void onBackPressed() {
        // Não faz nada
    }
}

File: Android Project/app/src/main/java/com/example/histoquiz/activities/MenuActivity.kt
package com.example.histoquiz.activities

import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat
import com.example.histoquiz.R
import com.example.histoquiz.databinding.ActivityMenuBinding
import com.google.firebase.auth.FirebaseAuth

/**
 * Classe utilizada para manipular os cliques recebidos na tela de menu do jogo, redirecionando o
 * usuário para as activities corretas relativas a cada interação
 */
class MenuActivity : AppCompatActivity(), View.OnClickListener {

    private lateinit var screen: ActivityMenuBinding

    /**
     * Método chamado assim que essa activity é invocada.
     * @param savedInstanceState - contém o estado anteriormente salvo da atividade (pode ser nulo)
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        screen = ActivityMenuBinding.inflate(layoutInflater)
        setContentView(screen.root)
        initGui()
        setInicialMenuOptions()
        hideSystemUI()
    }

    /**
     * Método chamado quando a janela atual da activity ganha ou perde o foco, é utilizado para es-
     * conder novamente a barra de status e a navigation bar.
     * @param hasFocus - booleano que indica se a janela desta atividade tem foco.
     */
    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        hideSystemUI()
    }

    /**
     * Método utilizado para fazer com que a barra de status e a navigation bar não sejam exibidas
     * na activity. Caso o usuário queira visualizá-las, ele deve realizar um movimento de arrastar
     * para cima (na navigation bar), ou para baixo (na status bar), o que fará com que elas apare-
     * çam por um momento e depois sumam novamente.
     */
    private fun hideSystemUI() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            WindowCompat.setDecorFitsSystemWindows(window, false)
            val controller = WindowCompat.getInsetsController(window, window.decorView)
            if (controller != null) {
                controller.hide(WindowInsetsCompat.Type.statusBars())
                controller.hide(WindowInsetsCompat.Type.navigationBars())
                controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            }
        } else {
            window.decorView.systemUiVisibility = (View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                    or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                    or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                    or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                    or View.SYSTEM_UI_FLAG_FULLSCREEN
                    or View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY)
        }
    }

    /** Método utilizado para obter uma referência para os quatro botões presentes na view
     * relacionada a essa activity
     */
    fun initGui() {
        screen.signOutTXT.tag = "SIGN_OUT"
        screen.signOutTXT.setOnClickListener(this)
        screen.button1.setOnClickListener(this)
        screen.button2.setOnClickListener(this)
        screen.button3.setOnClickListener(this)
        screen.button4.setOnClickListener(this)
        screen.goBackBTN.setOnClickListener(this)
    }

    /**
     * Método utilizado para redirecionar os cliques recebidos nos botões da view relacionada
     * a essa activity para o método que lidará corretamente com ele
     * @param v - view onde o clique ocorreu
     */
    override fun onClick(v: View) {
        when (v.tag as String) {
            "MINHA_CONTA" -> myAccount()
            "JOGAR", "LOCAL_GAME" -> setPlayOptions()
            "REVISAR" -> setReviewOptions()
            "SOBRE_O_JOGO" -> setAboutOptions()
            "JOGAR_PC" -> playAgainstPC()
            "JOGAR_LOCAL" -> playLocalGame()
            "CONVIDAR_AMIGO" -> inviteFriend()
            "O_QUE_E" -> whatItIs()
            "COMO_JOGAR" -> howToPlay()
            "FALE_CONOSCO" -> contactUs()
            "SISTEMA_REPRODUTOR" -> showReview("Sistema Reprodutor")
            "SISTEMA_DIGESTORIO" -> showReview("Sistema Digestório")
            "SISTEMA_CARDIOPULMONAR" -> showReview("Sistema Cardiopulmonar")
            "SISTEMA_OSTEOMUSCULAR" -> showReview("Sistema Osteomuscular")
            "MENU_INICIAL" -> setInicialMenuOptions()
            "CRIAR_SALA" -> createGameRoom()
            "ENTRAR_SALA" -> joinGameRoom()
            "SIGN_OUT" -> {
                FirebaseAuth.getInstance().signOut()
                val troca = Intent(this@MenuActivity, SignInActivity::class.java)
                startActivity(troca)
            }
        }
    }

    /**
     * Método utilizado para trocar para a activity responsável por exibir os
     * detalhes da conta do usuário
     */
    private fun myAccount() {
        val troca = Intent(this@MenuActivity, MyAccountActivity::class.java)
        startActivity(troca)
    }

    /**
     * Método que exibe ao jogador as opções de jogo disponíveis: jogar contra
     * o computador ou jogar online. Para isso, ele se aproveita dos botões do
     * menu: "jogar" vira "jogar online" e "revisar" vira "jogar contra o com-
     * putador". Os demais botões são desabilitados e ficam invisíveis
     */
    private fun setPlayOptions() {
        screen.button1.visibility = View.VISIBLE
        screen.button1.text = getString(R.string.local)
        screen.button1.tag = "JOGAR_LOCAL"
        screen.button1.isEnabled = true
        screen.button2.visibility = View.VISIBLE
        screen.button2.text
        screen.button2.text = getString(R.string.convidarAmigo)
        screen.button2.tag = "CONVIDAR_AMIGO"
        screen.button2.isEnabled = true
        screen.button3.visibility = View.VISIBLE
        screen.button3.text = getString(R.string.jogarPC)
        screen.button3.tag = "JOGAR_PC"
        screen.button3.isEnabled = true
        screen.button4.visibility = View.GONE
        screen.button4.isEnabled = false
        screen.goBackBTN.visibility = View.VISIBLE
        screen.goBackBTN.isEnabled = true
        screen.goBackBTN.tag = "MENU_INICIAL"
    }

    /**
     * Método utilizado para redirecionar dessa activity para a que exibe
     */
    private fun setReviewOptions() {
        screen.button1.text = getString(R.string.sistemaReprodutor)
        screen.button1.tag = "SISTEMA_REPRODUTOR"
        screen.button2.text = getString(R.string.sistemaDigestorio)
        screen.button2.tag = "SISTEMA_DIGESTORIO"
        screen.button3.text = getString(R.string.sistemaCardiopulmonar)
        screen.button3.tag = "SISTEMA_CARDIOPULMONAR"
        screen.button4.text = getString(R.string.sistemaOsteomuscular)
        screen.button4.tag = "SISTEMA_OSTEOMUSCULAR"
        screen.goBackBTN.visibility = View.VISIBLE
        screen.goBackBTN.isEnabled = true
        screen.goBackBTN.tag = "MENU_INICIAL"
    }

    /**
     * Método que exibe ao jogador as opções sobre o jogo: o que é, como jogar e fale co-
     * nosco. Para isso, ele se aproveita dos botões do menu: "jogar", transformando-o em
     * "O que é"; "revisar", transformando-o em "como jogar"; e "sobre o jogo", transfor-
     * mando-o em "fale conosco"
     */
    private fun setAboutOptions() {
        screen.button1.visibility = View.GONE
        screen.button2.text = getString(R.string.oQue)
        screen.button2.tag = "O_QUE_E"
        screen.button3.text = getString(R.string.comoJogar)
        screen.button3.tag = "COMO_JOGAR"
        screen.button4.text = getString(R.string.faleConosco)
        screen.button4.tag = "FALE_CONOSCO"
        screen.goBackBTN.visibility = View.VISIBLE
        screen.goBackBTN.isEnabled = true
        screen.goBackBTN.tag = "MENU_INICIAL"
    }

    /**
     * Método utilizado para redirecionar o jogador diretamente para a tela de jogo,
     * já que ele irá jogar contra o próprio celular
     */
    private fun playAgainstPC() {
        val troca = Intent(this@MenuActivity, GameActivity::class.java)
        troca.putExtra("matchCreator", true)
        troca.putExtra("PCopponent", true)
        troca.putExtra("opponentUID", "0")
        startActivityForResult(troca, 999)
    }

    /**
     * Método utilizado para exibir ao jogador as opções existentes relativas a um jogo local: criar
     * uma sala ou entrar em uma já existente
     */
    private fun playLocalGame() {
        screen.button1.visibility = View.GONE
        screen.button1.isEnabled = false
        screen.button2.text = getString(R.string.criarSala)
        screen.button2.tag = "CRIAR_SALA"
        screen.button3.text = getString(R.string.entrarSala)
        screen.button3.tag = "ENTRAR_SALA"
        screen.button4.visibility = View.GONE
        screen.button4.isEnabled = false
        screen.goBackBTN.visibility = View.VISIBLE
        screen.goBackBTN.isEnabled = true
        screen.goBackBTN.tag = "LOCAL_GAME"
    }

    /**
     * Método utilizado para redirecionar o jogador para a tela de configurar as métricas
     * necessárias para a criação de uma partida online
     */
    private fun createGameRoom() {
        val troca = Intent(this@MenuActivity, ConfigLocalGameActivity::class.java)
        startActivity(troca)
    }

    /**
     * Método utilizado para redirecionar o usuário a tela onde ele deve informar o código relati-
     * vo a sala do jogo no qual ele deseja entrar
     */
    private fun joinGameRoom() {
        val troca = Intent(this@MenuActivity, EnterLocalGameActivity::class.java)
        startActivity(troca)
    }

    /**
     * Método utilizado para redirecionar dessa activity para a que exibe a tela
     * utilizada para convidar amigo para jogar
     */
    private fun inviteFriend() {
        val troca = Intent(this@MenuActivity, InviteFriendToPlayActivity::class.java)
        startActivity(troca)
    }

    /**
     * Método utilizado para redirecionar dessa activity para a que descreve
     * o que o HistoQuiz é
     */
    private fun whatItIs() {
        val troca = Intent(this@MenuActivity, WhatItIsActivity::class.java)
        startActivity(troca)
    }

    /**
     * Método utilizado para redirecionar dessa activity para a que descreve
     * como jogar o HistoQuiz
     */
    private fun howToPlay() {
        val troca = Intent(this@MenuActivity, RulesActivity::class.java)
        startActivity(troca)
    }

    /**
     * Método utilizado para redirecionar dessa activity para a que permite
     * ao usuário enviar um email para os desenvolvedores do HistoQuiz
     */
    private fun contactUs() {
        val troca = Intent(this@MenuActivity, ContactActivity::class.java)
        startActivity(troca)
    }

    /**
     * Método utilizado para redirecionar dessa activity para a que exibe ao usuário
     * os dados para ele revisar sobre o sistema escolhido
     * @param kind - sistema do corpo humano que o usuário deseja revisar, a saber:
     * 1 - Sistema reprodutor
     * 2 - Sistema digestório
     * 3 - Sistema cardiopulmonar
     * 4 - Sistema osteomuscular
     */
    private fun showReview(kind: String) {
        val troca = Intent(this@MenuActivity, RevisionActivity::class.java)
        troca.putExtra("selectedSystem", kind)
        startActivity(troca)
    }

    /**
     * Método utilizado para setar tags para os botões presentes na view relacionada
     * a essa activity para a sua primeira execução, que exibe o menu principal
     */
    fun setInicialMenuOptions() {
        screen.button1.text = getString(R.string.minhaConta)
        screen.button1.tag = "MINHA_CONTA"
        screen.button1.visibility = View.VISIBLE
        screen.button1.isEnabled = true
        screen.button2.text = getString(R.string.jogar)
        screen.button2.tag = "JOGAR"
        screen.button2.visibility = View.VISIBLE
        screen.button2.isEnabled = true
        screen.button3.text = getString(R.string.revisar)
        screen.button3.tag = "REVISAR"
        screen.button3.visibility = View.VISIBLE
        screen.button3.isEnabled = true
        screen.button4.text = getString(R.string.sobreJogo)
        screen.button4.tag = "SOBRE_O_JOGO"
        screen.button4.visibility = View.VISIBLE
        screen.button4.isEnabled = true
        screen.goBackBTN.visibility = View.GONE
        screen.goBackBTN.isEnabled = false
    }

    /**
     * Sobrescreve o método que define o que deve ser feito quando o botão
     * pressionar do celular for pressionado. Como não quero que ele seja
     * utilizado, o método fica vazio
     */
    override fun onBackPressed() {
        // Não faz nada
    }
}